gax/src/main/java/com/google/api/gax/batching/BlockingSemaphore.java

```diff
@@ -43,31 +46,100 @@ private static void checkNotNegative(long l) {
   BlockingSemaphore(long permits) {
     checkNotNegative(permits);
     this.currentPermits = permits;
+    this.limit = permits;
   }
 
-  public synchronized void release(long permits) {
+  public void release(long permits) {
     checkNotNegative(permits);
-
-    currentPermits += permits;
-    notifyAll();
+    synchronized (updateLock) {
+      // If more permits are returned than what was originally set, we need to add these extra
+      // permits to the limit too
+      currentPermits += permits;
+      if (currentPermits > limit) {
+        limit = currentPermits;
+      }
+    }
+    synchronized (waitLock) {
+      waitLock.notifyAll();
+    }
   }
 
-  public synchronized boolean acquire(long permits) {
+  public boolean acquire(long permits) {
     checkNotNegative(permits);
 
     boolean interrupted = false;
-    while (currentPermits < permits) {
-      try {
-        wait();
-      } catch (InterruptedException e) {
-        interrupted = true;
+    while (!interrupted) {
+      synchronized (waitLock) {
+        if (currentPermits < permits) {
+          try {
+            waitLock.wait();
+          } catch (InterruptedException e) {
+            interrupted = true;
+          }
+        }
+      }
+      synchronized (updateLock) {
+        if (currentPermits < permits) {
+          continue;
+        }
+        currentPermits -= permits;
+        break;
       }
     }
-    currentPermits -= permits;
 
     if (interrupted) {
       Thread.currentThread().interrupt();
     }
     return true;
   }
+
+  public boolean acquirePartial(long permits) {
+    checkNotNegative(permits);
+
+    boolean interrupted = false;
+    while (!interrupted) {
```

Comment:
Please see the previous comment about interrupts. The same concerns applies here.